src/ArborX_BruteForce.hpp

```diff
@@ -0,0 +1,124 @@
+/****************************************************************************
+ * Copyright (c) 2012-2019 by the ArborX authors                            *
+ * All rights reserved.                                                     *
+ *                                                                          *
+ * This file is part of the ArborX library. ArborX is                       *
+ * distributed under a BSD 3-clause license. For the licensing terms see    *
+ * the LICENSE file in the top-level directory.                             *
+ *                                                                          *
+ * SPDX-License-Identifier: BSD-3-Clause                                    *
+ ****************************************************************************/
+
+#ifndef ARBORX_BRUTE_FORCE_HPP
+#define ARBORX_BRUTE_FORCE_HPP
+
+#include <ArborX_AccessTraits.hpp>
+#include <ArborX_Box.hpp>
+#include <ArborX_CrsGraphWrapper.hpp>
+#include <ArborX_DetailsBruteForceImpl.hpp>
+#include <ArborX_DetailsKokkosExtAccessibilityTraits.hpp>
+
+#include <Kokkos_Core.hpp>
+
+namespace ArborX
+{
+
+template <typename MemorySpace>
+class BruteForce
+{
+public:
+  using memory_space = MemorySpace;
+  static_assert(Kokkos::is_memory_space<MemorySpace>::value, "");
+  using size_type = typename MemorySpace::size_type;
+  using bounding_volume_type = Box;
+
+  BruteForce() = default;
+
+  template <typename ExecutionSpace, typename Primitives>
+  BruteForce(ExecutionSpace const &space, Primitives const &primitives);
+
+  KOKKOS_FUNCTION
+  size_type size() const noexcept { return _size; }
+
+  KOKKOS_FUNCTION
+  bool empty() const noexcept { return size() == 0; }
+
+  KOKKOS_FUNCTION
+  bounding_volume_type bounds() const noexcept { return _bounds; }
+
+  template <typename ExecutionSpace, typename Predicates, typename Callback,
+            typename Ignore = int>
+  void query(ExecutionSpace const &space, Predicates const &predicates,
+             Callback const &callback, Ignore = Ignore()) const;
+
+  template <typename ExecutionSpace, typename Predicates,
+            typename CallbackOrView, typename View, typename... Args>
+  std::enable_if_t<Kokkos::is_view<std::decay_t<View>>{}>
+  query(ExecutionSpace const &space, Predicates const &predicates,
+        CallbackOrView &&callback_or_view, View &&view, Args &&... args) const
+  {
+    ArborX::query(*this, space, predicates,
+                  std::forward<CallbackOrView>(callback_or_view),
+                  std::forward<View>(view), std::forward<Args>(args)...);
+  }
+
+private:
+  size_type _size;
+  bounding_volume_type _bounds;
+  Kokkos::View<bounding_volume_type *, memory_space> _bounding_volumes;
```

Comment:
This seems like a wrong approach. If a user's access traits return `Point`, there is no reason to store them as `Box`. The reason we do that in BVH right now is due to having the same type for internal and leaf nodes. We don't have this problem here.

Having said that, it's unclear to me how to properly handle this, as `Primitives` are only available in the constructor with the current interface, and this requires some compile time decision. Maybe a suitable workaround would be to template `BruteForce` on this type.

Reply:
Although these are points that I'd like to discuss in general as part of the bigger picture (semantics of our data structures and our access traits), the proposed implementation is the right one because it is consistent with the current design of the bounding volume hierarchy.  The changes you suggest raise a number of questions that are beyond the scope of this PR (e.g. template parameters of the data structure, requirements on the return type of `get()` from the primitives access type, etc.). Let's not get into these arguing some (unproven) early performance considerations.

Reply:
This comment is only to reply to your last comment:

> Let's not get into these arguing some (unproven) early performance considerations.

I tried to do a simple patch (won't work in general, but works with the benchmark): 
[Box2Point.patch.txt](https://github.com/arborx/ArborX/files/5859608/Box2Point.patch.txt)

The results confirm my expectation that it may be significantly faster to use `Point` instead of `Box` when primitives are points.
[Box.log](https://github.com/arborx/ArborX/files/5859614/Box.log)
[Point.log](https://github.com/arborx/ArborX/files/5859615/Point.log)
The experiments were run on my workstation, with sizes increasing from 100 to 4000. In general, serial construction was 1.5x fast, serial search 2x-3x+ times faster, Cuda construction unchanged, Cuda search 1-1.2x times faster.